kostyor/upgrades/engine.py

```diff
@@ -152,17 +152,21 @@ def start(self):
         # keystone, then with nova, and so on. See iteration details in
         # get_controllers() docstring.
         for host in get_controllers(hosts):
+            self.driver.pre_host_upgrade_hook(host)
             for service in iterservices(host):
+                self.driver.pre_service_upgrade_hook(service)
```

Comment:
Right. The pre and post hooks are not decorated with `abc.abstractmethod` - so a driver does not need to provide an implementation of them, like they do for the `*_upgrade` methods. So they can just rely on the pre and post hooks that they inherit from `UpgradeDriver` - which are noop